src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class MvcEndpointDataSource : EndpointDataSource
+    {
+        private readonly IActionDescriptorCollectionProvider _actions;
+        private readonly IActionInvokerFactory _invokerFactory;
+        private readonly IActionContextAccessor _actionContextAccessor;
+        private readonly List<Endpoint> _endpoints;
+
+        public MvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actions,
+            IActionInvokerFactory invokerFactory)
+            : this(actions, invokerFactory, actionContextAccessor: null)
+        {
+        }
+
+        public MvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actions,
+            IActionInvokerFactory invokerFactory,
+            IActionContextAccessor actionContextAccessor)
```

Comment:
It feels weird to pass `IActionContextAccessor` which is used for invoking to the data source

Reply:
Yeah I totes agree. We could consider putting this into the implementation of `IActionInvokerFactory` or creating a wrapper around it.

Reply:
> It feels weird to pass IActionContextAccessor which is used for invoking to the data source

We can resolve both IActionInvokerFactory and IActionContextAccessor from within the invoker delegate through context.RequestServices